### ISSUE TITLE
Why the prediction of evaluation function use gold target sequence in  CNN seq2seq? 

### ISSUE THREAD

**Issue by fake-warrior8** [bentrevett/pytorch-seq2seq#84]
In preceding seq2seq models based on LSTM, you set teacher forcing ratio as 0 to make the evaluation of validation and test set not use the gold target sequence as the input of decoder. However, it's very confusing that the CNN seq2seq model directly uses the gold target sequence as the input of decoder. Although you use  'sos' token to let the model just see the previous t-1 gold target tokens when predicting next t-th target tokens, it still uses too much gold information of  target sequence, which is similar with the operation of setting the teacher forcing ratio as 100% in the seq2seq models based on LSTM. (a cheating prediction)

I read the paper of CNN seq2seq model.
https://arxiv.org/abs/1705.03122
they use an incomprehensible saying:
We proceed similarly for output elements that were already generated by the decoder network to yield output element representations that are being fed back into the decoder network.
(It seems that it used an iterative embedding, but I don't know how it works and this paper didn't explain this.)

Maybe I know too little about seq2seq task of NLP, and I am really confused about this operation.

**Comment by fake-warrior8** [comment]
In line 4, it should be  'Although you use pad token', the 'pad' with angle brackets didn't show..

**Comment by bentrevett** [comment]
Yes, it's basically the same as using a teacher forcing ratio of 100%. 

From what I understand the model is designed to be trained with teacher forcing as the main advantage it has over traditional RNN based models for sequence-to-sequence tasks is that it can be massively parallelized. If you use the model to iteratively generate/decode one token at a time and then feed that back into the model then you lose this advantage. 

It is possible to train it with iterative decoding, this is exactly what we do in the `translate_sentence` function. It would be an interesting experiment to train the model using iterative decoding and see how fast and accurately it performs.

**Comment by fake-warrior8** [comment]
> Yes, it's basically the same as using a teacher forcing ratio of 100%.
> 
> From what I understand the model is designed to be trained with teacher forcing as the main advantage it has over traditional RNN based models for sequence-to-sequence tasks is that it can be massively parallelized. If you use the model to iteratively generate/decode one token at a time and then feed that back into the model then you lose this advantage.
> 
> It is possible to train it with iterative decoding, this is exactly what we do in the `translate_sentence` function. It would be an interesting experiment to train the model using iterative decoding and see how fast and accurately it performs.

Thanks a lot.
I thought your training for train set and evaluation for validate and test set were a complete part, and I didn't look the _traislate_sentence_ function before.
I find some other solutions to evaluate by iterative decoding in your evaluation function, which is suitable for batch evaluation. Firstly, I increased the trg length of every batch to the max_length by padding with 'pad'. Then I judge each batch predictions whether all get the 'eos' token, if so, I break the decoder loop. This will only incerase a little evaluation time.
